arch/djgpp/audio.c

```diff
@@ -0,0 +1,281 @@
+  return i > 0;
+}
+
+void init_audio_platform(struct config_info *conf)
+{
+  _go32_dpmi_seginfo new_irq_handler;
+  int buffer_size_bytes;
+  // Try to find a Sound Blaster
+  char *sb_env = getenv("BLASTER");
+  if(sb_env != NULL)
+    audio_sb_parse_env(&sb_cfg, sb_env);
+  if(sb_cfg.port != 0)
+    if(!audio_sb_dsp_detect())
+      sb_cfg.port = 0;
+  if(sb_cfg.port != 0)
+  {
+    audio_sb_dsp_write(0xE1); // version
+    sb_cfg.version_major = audio_sb_dsp_read();
+    sb_cfg.version_minor = audio_sb_dsp_read();
+
+    if(sb_cfg.version_major < 4)
+      sb_cfg.port = 0;
+    else if(sb_cfg.irq >= 8) // TODO: support IRQ8-15?
+      sb_cfg.port = 0;
+    else if(sb_cfg.dma16 < 5 || sb_cfg.dma16 >= 8)
+      sb_cfg.port = 0;
+  }
+
+  if(sb_cfg.port != 0)
+  {
+    // configure sample rate
+    if(audio.output_frequency < 5000)
+      audio.output_frequency = 5000;
+    else if(audio.output_frequency > 44100)
+      audio.output_frequency = 44100;
+
+    if(sb_cfg.version_major >= 4) // SB16
+    {
+      audio_sb_dsp_write(0x41); // set sampling rate
+      audio_sb_dsp_write(audio.output_frequency >> 8);
+      audio_sb_dsp_write(audio.output_frequency & 0xFF);
+    }
+    /* else // pre-SB16
+    {
+      Uint8 time_constant = 256 - (500000 / audio.output_frequency);
+      audio.output_frequency = 500000 / (256 - time_constant);
+      audio_sb_dsp_write(0x40); // set time constant
+      audio_sb_dsp_write(time_constant);
+    } */
+
+    // buffer_samples * 2 (stereo) * 2 (16-bit) * 2 (doubled)
+    // each sample is 4 bytes
+    if(audio.buffer_samples > 4096)
+      audio.buffer_samples = 4096;
+    else if(audio.buffer_samples <= 0)
+      audio.buffer_samples = 2048;
+
+    buffer_size_bytes = audio.buffer_samples << 3;
+    sb_cfg.nearptr_buffer_enabled = djgpp_push_enable_nearptr();
+
+    audio.mix_buffer = malloc(audio.buffer_samples << 3);
+
+    // allocate memory, without crossing 64K boundary, and clean it
+    sb_cfg.buffer_segment = djgpp_malloc_boundary(buffer_size_bytes, 65536, &sb_cfg.buffer_selector);
+    if(sb_cfg.nearptr_buffer_enabled)
+    {
+      sb_cfg.nearptr_buffer_mapping.address = sb_cfg.buffer_segment << 4;
+      sb_cfg.nearptr_buffer_mapping.size = buffer_size_bytes;
+      if(__dpmi_physical_address_mapping(&sb_cfg.nearptr_buffer_mapping) != 0)
+      {
+        sb_cfg.nearptr_buffer_enabled = false;
+        djgpp_pop_enable_nearptr();
+      }
+      else
+      {
+        sb_cfg.buffer = (Uint8*)(sb_cfg.nearptr_buffer_mapping.address + __djgpp_conventional_base);
+        memset(sb_cfg.buffer, 0, buffer_size_bytes);
+      }
+    }
+
+    if(!sb_cfg.nearptr_buffer_enabled)
+    {
+      sb_cfg.buffer = malloc(buffer_size_bytes);
+      memset(sb_cfg.buffer, 0, buffer_size_bytes);
+      dosmemput(sb_cfg.buffer, buffer_size_bytes, sb_cfg.buffer_segment << 4);
+    }
+
+    // lock C irq handler
+    // (TODO: rewrite handler in ASM?)
+    _go32_dpmi_lock_code(audio_sb_interrupt, 1024);
```

Comment:
It'd be nice to make this warning die but I'm not sure what short of adding a pragma to disable `-Wpedantic`.
```
  CC       arch/djgpp/audio.c
arch/djgpp/audio.c: In function 'init_audio_platform':
arch/djgpp/audio.c:224:26: warning: ISO C forbids passing argument 1 of '_go32_dpmi_lock_code' between function pointer and 'void *' [-Wpedantic]
  224 |     _go32_dpmi_lock_code(audio_sb_interrupt, 1024);
      |                          ^~~~~~~~~~~~~~~~~~
In file included from arch/djgpp/audio.c:26:
d:\djgpp\i586-pc-msdosdjgpp\sys-include\dpmi.h:337:33: note: expected 'void *' but argument is of type 'void (*)(void)'
  337 | int _go32_dpmi_lock_code( void *_lockaddr, unsigned long _locksize);
      |                           ~~~~~~^~~~~~~~~
```

Reply:
I would just leave it alone, the specific thing it does is a *very* ugly hack anyway...

Reply:
Looking [more into go32](http://www.delorie.com/djgpp/v2faq/faq18_10.html), yeah, I agree that this warning should stick around for now.